src/torrent.cpp

```diff
@@ -8920,11 +8918,13 @@ namespace libtorrent
 		// if we're in upload only mode and we're auto-managed
 		// leave upload mode every 10 minutes hoping that the error
 		// condition has been fixed
-		if (m_upload_mode && m_auto_managed
-			&& int(m_ses.session_time() - m_upload_mode_time)
-			>= settings().get_int(settings_pack::optimistic_disk_retry))
+		if (m_upload_mode && m_auto_managed)
 		{
-			set_upload_mode(false);
+			// cast ratio<1l, 1000000000l> to seconds_t ratio
+			std::int32_t upload_mode_duration = duration_cast<seconds_t>(
```

Comment:
maybe there should be a ``torrent::upload_mode_time()``, like there there is for seeding_time, active_time and finished_time. Then this expression would be a lot more intuitive, and it could probably be moved back into the if-condition